Parse paths as unicode strings or bytes

aabawer opened the pull request:
We would like to preserve Python 2 handle of paths as either unicode (ascii) strings or bytes.
We apply path converter for accepting PyObject and detecting if its either given as bytes or if its given as unicode. In the latter case we convert the unicode string to its bytes encoding and treat path internally as bytes throughout sanlock API. Tests are added and modified accordingly.

aabawer committed:
python: Add pypath_converter

In Python 3 we would like to handle paths as either unicode strings
or bytes. For this purpose we have a c-API called PyUnicode_FSConverter.
In this patch the missing API is implemented for Py2 based on Py3 [1]
and a unified no-cleanup behavior is enforced for Py3 path conversion as well.

[1] https://github.com/python/cpython/blob/master/Objects/unicodeobject.c#L3818

aabawer committed:
python: Apply pypath_converter to parse_resource util function

We should be able to parse resource paths as either unicode or bytes.

Stub tests involving valid arguments parsing of resources paths
are set with additional permutations with no xfails expected.

aabawer committed:
python: Apply pypath_converter to get_alignment API

We should be able to parse device path as either unicode or bytes.

New stub test was added for get_alignment API for path premutations
with no xfails expected.

aabawer committed:
python: Add pypath_converter to init_lockspace API

We should be able to parse lockspace path as either unicode or bytes.

Stub test for init_lockspace was set with additional permutations
over lockspace path with no expected xfails.

aabawer committed:
python: Apply pypath_converter to write_lockspace API

We should be able to parse lockspace path as either unicode or bytes.

Stub for write_lockspace test was set with additional permutations
over paths with no expected xfails.

aabawer committed:
python: Apply pypath_converter to read_lockspace API

We should be able to parse lockspace path as either unicode or bytes.

New stub test for read_lockspace API was added with path permutations
with no xfails expected.

aabawer committed:
python: Apply pypath_converter to read_resource API

We should be able to parse resource path as either unicode or bytes.

New stub test for read_resource API is added with path premutations
with no expected xfails.

aabawer committed:
python: Apply pypath_converter to add_lockspace

We should be able to parse lockspace path as either unicode or bytes.

New stub test is added for add_lockpace API with path permutation
with no xfails expected.

aabawer committed:
python: Apply pypath_converter to inq_lockspace API

We would like to parse lockspace path as either unicode or bytes.

Stub test for inq_lockspace API is set with additional permutation
for paths with no xfails expected.

aabawer committed:
python: Apply pypath_converter to rem_lockspace API

We would like to parse lockspace path as either unicode or bytes.

Stub test for rem_lockspace API is set with additional permutation
over path with no expected xfails.

aabawer committed:
python: Add pypath_converter to killpath API

We would like killpath to be able to handle paths as either unicode
and bytes, as well as the ones given in the command args.

Stub test for killpath API was modified to permutate over all paths
without expected xfails.

nirs commented on python/sanlock.c:
```diff
@@ -181,31 +182,33 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
 
-        strncpy(res->disks[i].path, path, SANLK_PATH_LEN - 1);
+        strncpy(res->disks[i].path, PyBytes_AsString(path), SANLK_PATH_LEN - 1);
+        Py_DECREF(path);
```
We should use Py_CLEAR(path) here to set the path to NULL immediately after decrefing it. The current code works, but if someone will remove line 188 setting the path to NULL, we will crash on parsing failure.

nirs commented on python/sanlock.c:
```diff
@@ -181,31 +182,33 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
     return 0;
 
 exit_fail:
+    Py_XDECREF(path);
```
Cleaning up path both globally and both once per loop is complicated and fragile. It is better to extract the code for parsing single (path, offset) item to a helper function that will hide the temporary objects  and have simple cleanup.

So should have something like:
```
parse_disks()
{
    allocate res...

    for (i = 0; i < num_disks; i++) {
        disk = PyList_GetItem(obj, i);
        if (parse_disks_item(disk, &res->disks[i]))
            goto exit_fail;
    }
    
exit_fail:
    ...
}
```

nirs reviewed (changes requested):
Reviewed up to python: Apply pypath_converter to parse_resource util function.

aabawer commented:
Updated the PR, Thanks.

On Tue, Jun 4, 2019 at 9:18 PM Nir Soffer <notifications@github.com> wrote:

> *@nirs* requested changes on this pull request.
>
> Reviewed up to python: Apply pypath_converter to parse_resource util
> function.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/19#discussion_r290391196>:
>
> >              /* Override the error since it confusing in this context. */
>              set_error(PyExc_ValueError, "Invalid disk %s", disk);
>              goto exit_fail;
>          }
>
> -        strncpy(res->disks[i].path, path, SANLK_PATH_LEN - 1);
> +        strncpy(res->disks[i].path, PyBytes_AsString(path), SANLK_PATH_LEN - 1);
> +        Py_DECREF(path);
>
> We should use Py_CLEAR(path) here to set the path to NULL immediately
> after decrefing it. The current code works, but if someone will remove line
> 188 setting the path to NULL, we will crash on parsing failure.
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/19#discussion_r290426205>:
>
> >          res->disks[i].offset = offset;
>      }
>
>      *res_ret = res;
>      return 0;
>
>  exit_fail:
> +    Py_XDECREF(path);
>
> Cleaning up path both globally and both once per loop is complicated and
> fragile. It is better to extract the code for parsing single (path, offset)
> item to a helper function that will hide the temporary objects and have
> simple cleanup.
>
> So should have something like:
>
> parse_disks()
> {
>     allocate res...
>
>     for (i = 0; i < num_disks; i++) {
>         disk = PyList_GetItem(obj, i);
>         if (parse_disks_item(disk, &res->disks[i]))
>             goto exit_fail;
>     }
>
> exit_fail:
>     ...
> }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/19?email_source=notifications&email_token=AL3BPNPFAJ6TWIBOHAKKCSDPY2WXJA5CNFSM4HS6X5J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2RRKQI#pullrequestreview-245568833>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNOYHNRY7FQ5JW2EGXLPY2WXJANCNFSM4HS6X5JQ>
> .
>

nirs commented on python/sanlock.c:
```diff
@@ -159,6 +159,33 @@ pystring_as_cstring(PyObject *obj)
+    Py_XDECREF(path);
+    return rv;
+}
+
```
Looks great, so much simpler now.

aabawer replied:
should be "finally" since we handle success as well

nirs commented on tests/python_test.py:
```diff
@@ -600,8 +617,10 @@ def test_init_lockspace_parse_args(no_sanlock_daemon, name):
+def test_init_resource_parse_args(no_sanlock_daemon, name, filename, encoding):
+    path = util.generate_path("/tmp/", filename, encoding)
+    disks = [(path, 0)]
     with raises_sanlock_errno(errno.ENOENT):
```
The tests changes should be separate patch. You are missing here improvement of the tests and changing the code. This is very likely to fail creating tests that do not work. I will have more confidence when I see the tests fail - before this change.

aabawer replied:
this is what we talked about before this PR - to change each test as part
of each relevant patch

aabawer replied:
unless you meant the case where the args parse test is added for the first time.

nirs replied:
We discussed that the patch fixing a failing test should enable the failing test, but here you modified the test and this make the code harder to review. If you think we need to modify the tests by adding parameters for path, why not submit this change as a separate commit before you modify the code?

nirs replied:
The only test change that should be included in this commit is replacing:

    @pytest.mark.parametrize("filename,encoding", FILE_NAMES)

With:

    @pytest.mark.parametrize("filename,encoding", FILE_NAMES_NO_XFAILS)

nirs reviewed (changes requested):
Reviewed again up to:
python: Apply pypath_converter to parse_resource util function

aabawer commented:
this is what we talked about before this PR - to change each test as part
of each relevant patch

On Tue, Jun 4, 2019 at 10:50 PM Nir Soffer <notifications@github.com> wrote:

> *@nirs* requested changes on this pull request.
>
> Reviewed again up to:
> python: Apply pypath_converter to parse_resource util function
> ------------------------------
>
> In python/sanlock.c
> <https://github.com/nirs/sanlock/pull/19#discussion_r290462625>:
>
> > +
> +    if (!PyArg_ParseTuple(disk, "O&K", pypath_converter, &path, &offset)) {
> +        /* Override the error since it confusing in this context. */
> +        set_error(PyExc_ValueError, "Invalid disk %s", disk);
> +        goto exit_fail;
> +    }
> +
> +    strncpy(res_disk->path, PyBytes_AsString(path), SANLK_PATH_LEN - 1);
> +    res_disk->offset = offset;
> +    rv = 1;
> +
> +exit_fail:
> +    Py_XDECREF(path);
> +    return rv;
> +}
> +
>
> Looks great, so much simpler now.
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/19#discussion_r290467871>:
>
> > @@ -600,8 +617,10 @@ def test_init_lockspace_parse_args(no_sanlock_daemon, name):
>
>
>  @pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
> -def test_init_resource_parse_args(no_sanlock_daemon, name):
> -    disks = [("path", 0)]
> +@pytest.mark.parametrize("filename,encoding", FILE_NAMES_NO_XFAILS)
> +def test_init_resource_parse_args(no_sanlock_daemon, name, filename, encoding):
> +    path = util.generate_path("/tmp/", filename, encoding)
> +    disks = [(path, 0)]
>      with raises_sanlock_errno(errno.ENOENT):
>
> The tests changes should be separate patch. You are missing here
> improvement of the tests and changing the code. This is very likely to fail
> creating tests that do not work. I will have more confidence when I see the
> tests fail - before this change.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/19?email_source=notifications&email_token=AL3BPNORXAU4MCLNHBXSSPTPY3BRHA5CNFSM4HS6X5J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2SHJHQ#pullrequestreview-245658782>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNKZOWFEN4F4VACITN3PY3BRHANCNFSM4HS6X5JQ>
> .
>

aabawer commented:
i think we are talking about the same stuff.

On Wed, Jun 5, 2019 at 12:28 AM Nir Soffer <notifications@github.com> wrote:

> *@nirs* commented on this pull request.
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/19#discussion_r290504680>:
>
> > @@ -600,8 +617,10 @@ def test_init_lockspace_parse_args(no_sanlock_daemon, name):
>
>
>  @pytest.mark.parametrize("name", LOCKSPACE_OR_RESOURCE_NAMES)
> -def test_init_resource_parse_args(no_sanlock_daemon, name):
> -    disks = [("path", 0)]
> +@pytest.mark.parametrize("filename,encoding", FILE_NAMES_NO_XFAILS)
> +def test_init_resource_parse_args(no_sanlock_daemon, name, filename, encoding):
> +    path = util.generate_path("/tmp/", filename, encoding)
> +    disks = [(path, 0)]
>      with raises_sanlock_errno(errno.ENOENT):
>
> The only test change that should be included in this commit is replacing:
>
> @pytest.mark.parametrize("filename,encoding", FILE_NAMES)
>
> With:
>
> @pytest.mark.parametrize("filename,encoding", FILE_NAMES_NO_XFAILS)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/19?email_source=notifications&email_token=AL3BPNM22FPPUPK2E5G3ZVTPY3M6VA5CNFSM4HS6X5J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2SUQSQ#discussion_r290504680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNO7ZAAYRFBLP27J3JDPY3M6VANCNFSM4HS6X5JQ>
> .
>

aabawer committed:
tests: Add path conversion tests to stub tests section

We add tests for path permutations involving xfail secnarios
to stub argument parsing tests. Also adding missing API tests.
Once we apply all patches involving the new path converter, we
expect all of them to pass with no fails in both python versions.

aabawer committed:
tests: Remove xfail marks from path file names

At this point we expect all API calls to handle file paths
as either unicode strings or bytes. This applies to both python
versions.

XPASS/xfail expected for test_write_resource_4k_invalid_sector_size.

aabawer commented:
Pushed update to PR for review.

nirs reviewed (commented):
Thanks, looks good.
